api/client/webclient.go

```diff
@@ -0,0 +1,188 @@
+	}
+
+	// Skip https cert verification, print a warning that this is insecure.
+	return NewInsecureWebClient(), nil
+}
+
+func NewInsecureWebClient() *http.Client {
+	return &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		},
+	}
+}
+
+// Ping serves two purposes. The first is to validate the HTTP endpoint of a
+// Teleport proxy. This leads to better user experience: users get connection
+// errors before being asked for passwords. The second is to return the form
+// of authentication that the server supports. This also leads to better user
+// experience: users only get prompted for the type of authentication the server supports.
+func Ping(ctx context.Context, proxyAddr string) (*PingResponse, error) {
+	clt, err := initClient(proxyAddr)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	resp, err := clt.Get("https://" + proxyAddr + "/webapi/ping")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	defer resp.Body.Close()
+	pr := &PingResponse{}
+	if err := json.NewDecoder(resp.Body).Decode(pr); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return pr, nil
+}
+
+// PingResponse contains data about the Teleport server like supported
+// authentication types, server version, etc.
+type PingResponse struct {
+	// Auth contains the forms of authentication the auth server supports.
+	Auth AuthenticationSettings `json:"auth"`
+	// Proxy contains the proxy settings.
+	Proxy ProxySettings `json:"proxy"`
+	// ServerVersion is the version of Teleport that is running.
+	ServerVersion string `json:"server_version"`
+	// MinClientVersion is the minimum client version required by the server.
+	MinClientVersion string `json:"min_client_version"`
+}
+
+// ProxySettings contains basic information about proxy settings
+type ProxySettings struct {
+	// Kube is a kubernetes specific proxy section
+	Kube KubeProxySettings `json:"kube"`
+	// SSH is SSH specific proxy settings
+	SSH SSHProxySettings `json:"ssh"`
+	// DB contains database access specific proxy settings
+	DB DBProxySettings `json:"db"`
+}
+
+// KubeProxySettings is kubernetes proxy settings
+type KubeProxySettings struct {
+	// Enabled is true when kubernetes proxy is enabled
+	Enabled bool `json:"enabled,omitempty"`
+	// PublicAddr is a kubernetes proxy public address if set
+	PublicAddr string `json:"public_addr,omitempty"`
+	// ListenAddr is the address that the kubernetes proxy is listening for
+	// connections on.
+	ListenAddr string `json:"listen_addr,omitempty"`
+}
+
+// SSHProxySettings is SSH specific proxy settings.
+type SSHProxySettings struct {
+	// ListenAddr is the address that the SSH proxy is listening for
+	// connections on.
+	ListenAddr string `json:"listen_addr,omitempty"`
+
+	// TunnelListenAddr
+	TunnelListenAddr string `json:"tunnel_listen_addr,omitempty"`
+
+	// PublicAddr is the public address of the HTTP proxy.
+	PublicAddr string `json:"public_addr,omitempty"`
```

Comment:
Not sure if this is the right time to resolve this as it's so entrenched in the codebase but this isn't really "an SSH specific setting". It should at least be renamed to `HTTPPublicAddr` or `WebPublicAddr`.

Reply:
Since this is part of the response for the pre-existing `Ping` endpoint, I don't think this is the right PR to make any changes to it. Additional considerations for are necessary.